homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,806 @@
+    requires_auth = False
+
+    def __init__(self, hass):
+        """Initializer."""
+        self.hass = hass
+        self._registered_services = {}
+        self._backlog = {}
+
+    def register(self, router):
+        """Register the view with a router."""
+        handler = request_handler_factory(self, self.async_notify)
+        router.add_route('notify', UpnpNotifyView.url, handler)
+
+    async def async_notify(self, request):
+        """Callback method for NOTIFY requests."""
+        if 'SID' not in request.headers:
+            return aiohttp.web.Response(status=412)
+
+        headers = request.headers
+        sid = headers['SID']
+        body = await request.text()
+
+        # find UpnpService by SID
+        if sid not in self._registered_services:
+            _LOGGER.debug('Storing NOTIFY in backlog for SID: %s', sid)
+            self._backlog[sid] = {'headers': headers, 'body': body}
+            return aiohttp.web.Response(status=202)
+
+        service = self._registered_services[sid]
+        status = service.on_notify(headers, body)
+        return aiohttp.web.Response(status=status)
+
+    @property
+    def callback_url(self):
+        """Full URL to be called by device/service."""
+        base_url = self.hass.config.api.base_url
+        return urllib.parse.urljoin(base_url, self.url)
+
+    def register_service(self, sid, service):
+        """Register a UpnpService under SID."""
+        if sid in self._registered_services:
+            raise RuntimeError('SID {} already registered.'.format(sid))
+
+        self._registered_services[sid] = service
+
+        if sid in self._backlog:
+            item = self._backlog[sid]
+            _LOGGER.debug('Re-playing backlogged NOTIFY for SID: %s', sid)
+            service.on_notify(item['headers'], item['body'])
+            del self._backlog[sid]
+
+    def unregister_service(self, sid):
+        """Unregister service by SID."""
+        if sid in self._registered_services:
+            del self._registered_services[sid]
+
+
+class HassUpnpRequester:
+    """async_upnp_client.UpnpRequester for home-assistant."""
+
+    def __init__(self, hass):
+        """Initializer."""
+        self.hass = hass
+
+    async def async_http_request(self, method, url, headers=None, body=None):
+        """Do a HTTP request."""
+        soap_action = headers['SOAPAction'] \
+            if 'SOAPAction' in (headers or []) else ''
+        _LOGGER.debug('%s.async_http_request(): method: %s, soap_action: %s',
+                      self, method, soap_action)
+        session = async_get_clientsession(self.hass)
+        with async_timeout.timeout(5, loop=self.hass.loop):
+            response = await session.request(method,
+                                             url,
+                                             headers=headers,
+                                             data=body)
+            response_body = await response.text()
+
+        await asyncio.sleep(0.25)
```

Comment:
I think the problem w/o sleeping here is caused by the transport closing (the old comments have disappeared, so just commenting here). With pure aiohttp I have usually used the context manager which should avoid such cases, but I don't know how this should work with the homeassistant helper.. @balloob do you have an idea here?